### ISSUE TITLE
#12639 seem to create random errors

### ISSUE THREAD

**Issue by Carreau** [ipython/ipython#12692]
@jetpacktuxedo your PR seem to randomly create errors like this:

https://travis-ci.com/github/ipython/ipython/jobs/423309288:

```
        if args.raise_error and p.returncode!=0:
>           raise CalledProcessError(p.returncode, cell)
E           subprocess.CalledProcessError: <unprintable CalledProcessError object>
```

As you wrote it maybe you will find this issue faster than I.

**Comment by jetpacktuxedo** [comment]
That's interesting... I'm not sure why that would be unprintable. I think `p.returncode` should be either `None` (if the process hasn't completed yet) or an integer return code, which would leave `cell` to make it unprintable, but I'm pretty sure that's just an encoded binary string.

It's possible that I introduced some sort of timing issue in the tests where it is trying to evaluate the success of the test before the test script call is done executing?

**Comment by Carreau** [comment]
It's possible, it's possible it's also a but in asyncio on a specific python version. I haven't really had the time to investigate; 
And I think the test is not expecting a non-zero return value; so it's likely possible something else is wrong.

**Comment by takluyver** [comment]
I think this occurs when `p.returncode` is None.

The `<unprintable ... object>` message appears when `str(x)` fails in the traceback module. Looking at the code of `CalledProcessError.__str__`, it is clearly assuming that it will always get an integer, and the string formatting with throw an error with None.

On Github actions, this seems to fail consistently rather than occasionally, so I'll figure out some workaround.

**Comment by takluyver** [comment]
The workaround/fix I've got now is in two parts:

- Wait for the process to finish in `_stream_communicate()`. The stdout & stderr pipes being closed doesn't automatically mean that the child process has terminated, or that we've processed its return code. The `.communicate()` methods in subprocess and asyncio both do this.
- If we reach the point of raising the error and returncode is still None, use -9 (i.e. SIGKILL) instead. I think this should only be possible when the process has been killed but we haven't got the return code yet.

I'm not sure about the logic in the KeyboardInterrupt handler, though - it's manipulating an asyncio process object, but outside the asyncio event loop, calling `time.sleep()` to wait between SIGINT/SIGTERM/SIGKILL. I'm not convinced `p.returncode` will actually get set even if the process exits correctly.
